File: frontend-react/packages/formula-engine/src/reference-transform-domain.ts
/** Canonical worksheet coordinate bounds and axis-reference mapping semantics. */
export const MAX_ROW_INDEX = 1_048_575;
export const MAX_COLUMN_INDEX = 16_383;

export interface StructuralShift {
  axis: 'row' | 'column';
  at: number;
  count: number;
  op: 'insert' | 'delete';
}

export interface CellShiftSelection {
  readonly startRow: number;
  readonly endRow: number;
  readonly startColumn: number;
  readonly endColumn: number;
}

export type CellShiftPointTransformResult =
  | { readonly kind: 'mapped'; readonly row: number; readonly column: number }
  | { readonly kind: 'deleted' }
  | { readonly kind: 'out-of-bounds'; readonly row: number; readonly column: number };

export type PointTransformResult =
  | { readonly kind: 'mapped'; readonly position: number }
  | { readonly kind: 'deleted' }
  | { readonly kind: 'out-of-bounds'; readonly position: number };

export type IntervalTransformResult =
  | { readonly kind: 'mapped'; readonly start: number; readonly end: number }
  | { readonly kind: 'deleted' }
  | { readonly kind: 'out-of-bounds'; readonly start: number; readonly end: number };

export class ReferenceTransformDomain {
  static mapCellShiftIndex(position: number, start: number, end: number, op: 'insert' | 'delete', maximum: number): number {
    if (!Number.isSafeInteger(position) || position < 0
      || !Number.isSafeInteger(start) || start < 0
      || !Number.isSafeInteger(end) || end < start
      || !Number.isSafeInteger(maximum) || maximum < 0 || end > maximum
      || (op !== 'insert' && op !== 'delete')) {
      throw new Error('Reference transform cell-shift index inputs are invalid');
    }
    if (position > maximum) return position;
    if (op === 'delete' && position >= start && position <= end) return -1;
    if (position < start) return position;
    return position + (op === 'insert' ? 1 : -1) * (end - start + 1);
  }

  static mapCellShiftPoint(
    row: number,
    column: number,
    selection: CellShiftSelection,
    axis: 'row' | 'column',
    op: 'insert' | 'delete',
  ): CellShiftPointTransformResult {
    if (!Number.isSafeInteger(row) || row < 0
      || !Number.isSafeInteger(column) || column < 0
      || !selection
      || !Number.isSafeInteger(selection.startRow) || selection.startRow < 0
      || !Number.isSafeInteger(selection.endRow) || selection.endRow < selection.startRow || selection.endRow > MAX_ROW_INDEX
      || !Number.isSafeInteger(selection.startColumn) || selection.startColumn < 0
      || !Number.isSafeInteger(selection.endColumn) || selection.endColumn < selection.startColumn || selection.endColumn > MAX_COLUMN_INDEX
      || (axis !== 'row' && axis !== 'column')
      || (op !== 'insert' && op !== 'delete')) {
      throw new Error('Reference transform cell-shift inputs are invalid');
    }
    const maximumRow = MAX_ROW_INDEX;
    const maximumColumn = MAX_COLUMN_INDEX;
    if (row > maximumRow || column > maximumColumn) return { kind: 'out-of-bounds', row, column };

    const start = axis === 'row' ? selection.startRow : selection.startColumn;
    const end = axis === 'row' ? selection.endRow : selection.endColumn;
    const position = axis === 'row' ? row : column;
    const inBand = axis === 'row'
      ? column >= selection.startColumn && column <= selection.endColumn && row >= start
      : row >= selection.startRow && row <= selection.endRow && column >= start;
    if (!inBand) return { kind: 'mapped', row, column };
    const maximum = axis === 'row' ? maximumRow : maximumColumn;
    const mappedPosition = this.mapCellShiftIndex(position, start, end, op, maximum);
    if (mappedPosition === -1) return { kind: 'deleted' };
    if (mappedPosition > maximum) {
      return {
        kind: 'out-of-bounds',
        row: axis === 'row' ? mappedPosition : row,
        column: axis === 'column' ? mappedPosition : column,
      };
    }
    return axis === 'row'
      ? { kind: 'mapped', row: mappedPosition, column }
      : { kind: 'mapped', row, column: mappedPosition };
  }

  static mapPoint(position: number, at: number, count: number, direction: 1 | -1, maximum: number): PointTransformResult {
    if (!Number.isSafeInteger(position) || position < 0
      || !Number.isSafeInteger(at) || at < 0
      || !Number.isSafeInteger(count) || count < 1
      || (direction !== 1 && direction !== -1)
      || !Number.isSafeInteger(maximum) || maximum < 0 || at > maximum || count > maximum + 1) {
      throw new Error('Reference transform point inputs are invalid');
    }
    if (position > maximum) return { kind: 'out-of-bounds', position };
    if (direction === -1 && position >= at && position - at < count) return { kind: 'deleted' };
    const mapped = direction === 1
      ? position >= at ? position + count : position
      : position >= at ? position - count : position;
    return Number.isSafeInteger(mapped) && mapped <= maximum
      ? { kind: 'mapped', position: mapped }
      : { kind: 'out-of-bounds', position: mapped };
  }

  static mapInterval(
    start: number,
    end: number,
    shift: StructuralShift,
  ): IntervalTransformResult {
    if (!Number.isSafeInteger(start) || start < 0 || !Number.isSafeInteger(end) || end < 0
      || !shift || (shift.axis !== 'row' && shift.axis !== 'column')
      || !Number.isSafeInteger(shift.at) || shift.at < 0
      || !Number.isSafeInteger(shift.count) || shift.count < 1
      || (shift.op !== 'insert' && shift.op !== 'delete')) {
      throw new Error('Reference transform interval inputs are invalid');
    }
    const maximum = shift.axis === 'row' ? MAX_ROW_INDEX : MAX_COLUMN_INDEX;
    if (start > maximum || end > maximum || shift.at > maximum || shift.count > maximum + 1
      || (shift.op === 'delete' && shift.count - 1 > maximum - shift.at)) {
      throw new Error('Reference transform interval inputs exceed worksheet bounds');
    }
    const low = Math.min(start, end);
    const high = Math.max(start, end);
    let interval: { readonly start: number; readonly end: number } | undefined;
    if (shift.op === 'insert') {
      if (shift.at <= low) interval = { start: low + shift.count, end: high + shift.count };
      else if (shift.at <= high) interval = { start: low, end: high + shift.count };
      else interval = { start: low, end: high };
    } else {
      const deletedEnd = shift.at + shift.count - 1;
      if (high < shift.at) interval = { start: low, end: high };
      else if (low > deletedEnd) interval = { start: low - shift.count, end: high - shift.count };
      else {
        const nextStart = low < shift.at ? low : shift.at;
        const nextEnd = high > deletedEnd ? high - shift.count : shift.at - 1;
        if (nextStart <= nextEnd) interval = { start: nextStart, end: nextEnd };
      }
    }
    if (!interval) return { kind: 'deleted' };
    return Number.isSafeInteger(interval.start) && Number.isSafeInteger(interval.end)
      && interval.start >= 0 && interval.end <= maximum
      ? { kind: 'mapped', start: interval.start, end: interval.end }
      : { kind: 'out-of-bounds', start: interval.start, end: interval.end };
  }
}

File: backend/src/main/java/com/xc/luckysheet/server/mutation/ReferenceTransformDomain.java
package com.xc.luckysheet.server.mutation;

/** Canonical server-side point and inclusive-interval mapping for worksheet references. */
final class ReferenceTransformDomain {
    static final int MAX_ROW_INDEX = 1_048_575;
    static final int MAX_COLUMN_INDEX = 16_383;

    enum CellAxis { ROW, COLUMN }

    enum Operation { INSERT, DELETE }

    enum CellPointKind { MAPPED, DELETED, OUT_OF_BOUNDS }

    record CellPointMapping(CellPointKind kind, Integer row, Integer column) { }

    enum PointKind { MAPPED, DELETED, OUT_OF_BOUNDS }

    record PointMapping(PointKind kind, Long position) {
    }

    enum IntervalKind { MAPPED, DELETED, OUT_OF_BOUNDS }

    record IntervalMapping(IntervalKind kind, Long start, Long end) {
    }

    private ReferenceTransformDomain() {
    }

    static long mapCellShiftIndex(int position, int start, int end, Operation operation, int maximum) {
        if (position < 0 || start < 0 || end < start || end > maximum || operation == null || maximum < 0) {
            throw new IllegalArgumentException("Reference transform cell-shift index inputs are invalid");
        }
        if (position > maximum) return position;
        if (operation == Operation.DELETE && position >= start && position <= end) return -1L;
        if (position < start) return position;
        long count = (long) end - start + 1;
        return position + (operation == Operation.INSERT ? count : -count);
    }

    static CellPointMapping mapCellShiftPoint(
            int row,
            int column,
            int startRow,
            int endRow,
            int startColumn,
            int endColumn,
            CellAxis axis,
            Operation operation
    ) {
        if (row < 0 || column < 0 || startRow < 0 || endRow < startRow || endRow > MAX_ROW_INDEX
                || startColumn < 0 || endColumn < startColumn || endColumn > MAX_COLUMN_INDEX
                || axis == null || operation == null) {
            throw new IllegalArgumentException("Reference transform cell-shift inputs are invalid");
        }
        if (row > MAX_ROW_INDEX || column > MAX_COLUMN_INDEX) {
            return new CellPointMapping(CellPointKind.OUT_OF_BOUNDS, row, column);
        }

        int start = axis == CellAxis.ROW ? startRow : startColumn;
        int end = axis == CellAxis.ROW ? endRow : endColumn;
        int position = axis == CellAxis.ROW ? row : column;
        boolean inBand = axis == CellAxis.ROW
                ? column >= startColumn && column <= endColumn && row >= start
                : row >= startRow && row <= endRow && column >= start;
        if (!inBand) return new CellPointMapping(CellPointKind.MAPPED, row, column);
        int maximum = axis == CellAxis.ROW ? MAX_ROW_INDEX : MAX_COLUMN_INDEX;
        long mappedPosition = mapCellShiftIndex(position, start, end, operation, maximum);
        if (mappedPosition == -1) return new CellPointMapping(CellPointKind.DELETED, null, null);
        if (mappedPosition > maximum) {
            return new CellPointMapping(CellPointKind.OUT_OF_BOUNDS,
                    axis == CellAxis.ROW ? Math.toIntExact(mappedPosition) : row,
                    axis == CellAxis.COLUMN ? Math.toIntExact(mappedPosition) : column);
        }
        return axis == CellAxis.ROW
                ? new CellPointMapping(CellPointKind.MAPPED, (int) mappedPosition, column)
                : new CellPointMapping(CellPointKind.MAPPED, row, (int) mappedPosition);
    }

    static PointMapping mapPoint(int position, int at, int count, boolean insert, int maximum) {
        if (position < 0 || at < 0 || count < 1 || maximum < 0 || at > maximum || (long) count > (long) maximum + 1) {
            throw new IllegalArgumentException("Reference transform point inputs are invalid");
        }
        if (position > maximum) return new PointMapping(PointKind.OUT_OF_BOUNDS, (long) position);
        if (!insert && position >= at && (long) position - at < count) {
            return new PointMapping(PointKind.DELETED, null);
        }
        long mapped = insert
                ? position < at ? position : (long) position + count
                : position < at ? position : (long) position - count;
        return mapped <= maximum
                ? new PointMapping(PointKind.MAPPED, mapped)
                : new PointMapping(PointKind.OUT_OF_BOUNDS, mapped);
    }

    static IntervalMapping mapInterval(int start, int end, int at, int count, boolean insert, int maximum) {
        if (start < 0 || end < 0 || at < 0 || count < 1 || maximum < 0
                || start > maximum || end > maximum || at > maximum || (long) count > (long) maximum + 1
                || (!insert && (long) count - 1 > (long) maximum - at)) {
            throw new IllegalArgumentException("Reference transform interval inputs are invalid");
        }
        long low = Math.min(start, end);
        long high = Math.max(start, end);
        long nextStart;
        long nextEnd;
        if (insert) {
            if (at <= low) {
                nextStart = low + count;
                nextEnd = high + count;
            } else if (at <= high) {
                nextStart = low;
                nextEnd = high + count;
            } else {
                nextStart = low;
                nextEnd = high;
            }
        } else {
            long deletedEnd = (long) at + count - 1;
            if (high < at) {
                nextStart = low;
                nextEnd = high;
            } else if (low > deletedEnd) {
                nextStart = low - count;
                nextEnd = high - count;
            } else {
                nextStart = low < at ? low : at;
                nextEnd = high > deletedEnd ? high - count : (long) at - 1;
            }
        }
        if (nextStart > nextEnd) return new IntervalMapping(IntervalKind.DELETED, null, null);
        if (nextStart < 0 || nextEnd > maximum) {
            return new IntervalMapping(IntervalKind.OUT_OF_BOUNDS, nextStart, nextEnd);
        }
        return new IntervalMapping(IntervalKind.MAPPED, nextStart, nextEnd);
    }
}
